Public/Src/Cache/ContentStore/Library/Stores/FileSystemContentStoreInternal.cs

```diff
@@ -349,6 +354,195 @@ private Task Upgrade2To3(Context context)
             return upgradeCacheBlobAction.PostAllAndComplete(EnumerateBlobPathsFromDisk());
         }
 
+        /// <summary>
+        /// Checks that the content on disk is correct and every file in content directory matches it's hash.
+        /// </summary>
+        /// <returns></returns>
+        public async Task<Result<SelfCheckResult>> SelfCheckAsync(Context context, CancellationToken token)
+        {
+            using (var trackingContext = TrackShutdown(context, token))
+            {
+                var result = await trackingContext.Context.PerformOperationAsync(
+                    Tracer,
+                    () =>
+                    {
+                        if (SelfCheckUpToDate())
+                        {
+                            return Task.FromResult(Result.Success(SelfCheckResult.UpToDate()));
+                        }
+
+                        return SelfCheckCoreAsync(trackingContext.Context);
+                    });
+
+                if (result)
+                {
+                    MarkSelfChecked(success: true);
+                }
+
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// A result of <see cref="SelfCheckAsync"/>.
+        /// </summary>
+        public struct SelfCheckResult
+        {
+            /// <summary>
+            /// Total number of invalid files.
+            /// </summary>
+            public int InvalidFiles { get; }
+
+            /// <summary>
+            /// Total number of files scanned on disk.
+            /// </summary>
+            public int TotalFilesScanned { get; }
+
+            /// <nodoc />
+            public SelfCheckResult(int invalidHashes, int totalFilesScanned) =>
+                (InvalidFiles, TotalFilesScanned) = (invalidHashes, totalFilesScanned);
+
+            /// <nodoc />
+            public static SelfCheckResult UpToDate() => new SelfCheckResult(invalidHashes: 0, totalFilesScanned: -1);
+
+            /// <inheritdoc />
+            public override string ToString() => $"InvalidHashes={InvalidFiles}, TotalFilesScanned={TotalFilesScanned}";
+        }
+
+        private async Task<Result<SelfCheckResult>> SelfCheckCoreAsync(OperationContext context)
+        {
+            var stopwatch = Stopwatch.StartNew();
+            var contentHashes = ReadContentHashesFromDisk(context);

```

Comment:
Prefer to use in-memory enumeration. Reduces I/O, and if a file isn't in the in-memory representation, it won't be pinned/placed anyway. Consider this issue of higher important than others.